Change plugin to use built in AWS provider.

gmetzker opened the pull request:
Issues:  #7 #4 
- Modified the plugin to use the built in AWS provider `serverless.getProvider('aws')` so region and account are pulled directly from the same context as serverless.  This is similar to the built-in plugins.

- Removed dependencies on aws-sdk and q

Issues: #5 
- Updated Lambda event permissions to have the same conditions that serverless typically creates.  This also fixes the issue where the event does not display on the Lambda console under Triggers.


## Tests
Added tests for the following areas:

- subscribeFunction
- unsubscribeFunction
- addEventPermission

gmetzker committed:
- Changed plugin to use built-in serverless provider.request for AWS SDK calls.
- Removed dependency on AWS SDK
- Added SourceArn so Lambda to AWS permission so it's more secure. This will also allow it to display in the Triggers tab in the Lambda console.
- Added additional Unit Tests for better coverage.

gmetzker commented:
Eh...eslint doesn't like the arrow functions () => { }.  
serverless was using these quite extensively so I figured they would be fine.

charlesbalderson commented:
@gmetzker 👍 This looks brilliant.

Govinda-Fichtner commented:
@gmetzker Any ideas when this will be merged?

massimocode commented:
@gmetzker I just wanted to say that we were having the same issues as described here: https://github.com/silvermine/serverless-plugin-external-sns-events/issues/7

Using the index.js from this PR enabled us to deploy our lambdas without issue. It would be awesome to get this merged and published to NPM so we can use it.

massimocode commented:
@jthomerson any chance you could review this PR?

jthomerson commented on package.json:
```diff
@@ -39,6 +37,7 @@
     "mocha-lcov-reporter": "1.2.0",
-    "sinon": "1.17.5"
+    "sinon": "1.17.5",
+    "bluebird": "^3.4.0"
```
Why was a dependency added on `bluebird` here? All projects in our `silvermine` group use `q` for promises, and we're not looking to switch. The only place I see you using Bluebird is in the tests, so that would need to be refactored.

jthomerson replied:
Additionally, as you notice on the rest of the dependencies, we do not use fuzzy matching for dependency versions. You would need to remove the `^` if you were to add a dependency.

gmetzker replied:
I used `bluebird` because the serverless source code and most plugin developers use this.  In any case I understand if you have preferences.

I've refactored to use `Q`, and updated the `devDependencies` to use `"q": "1.4.1"`

jthomerson commented on src/index.js:
```diff
@@ -21,9 +20,11 @@ module.exports = Class.extend({
             lifecycleEvents: [ 'subscribe' ],
          },
          unsubscribeExternalSNS: {
+            usage: 'Removes subscriptions to any SNS Topics defined by externalSNS.',
```
Please change `to` to `from`

gmetzker replied:
done

jthomerson commented on src/index.js:
```diff
@@ -35,6 +36,7 @@ module.exports = Class.extend({
       _.each(this._serverless.service.functions, function(fnDef, fnName) {
          _.each(fnDef.events, function(evt) {
             if (evt.externalSNS) {
+
```
erroneous whitespace that needs to be removed

gmetzker replied:
done

jthomerson commented on src/index.js:
```diff
@@ -54,67 +56,80 @@ module.exports = Class.extend({
+            params = {
+               TopicArn: info.TopicArn,
+               Protocol: 'lambda',
+               Endpoint: info.FunctionArn
```
Common practice in our codebases is to leave the trailing comma on the last items in arrays / objects. This eliminates the review noise if in a future review another line is added because adding a line doesn't then require a comma be added to the existing line.

gmetzker replied:
done

jthomerson commented on src/index.js:
```diff
@@ -54,67 +56,80 @@ module.exports = Class.extend({
-                     ' (deleted ' + info.SubscriptionArn + ')'
+                  self._serverless.cli.log(
+                    'Function ' + info.FunctionArn + ' is no longer subscribed to ' + info.TopicArn +
+                    ' (deleted ' + info.SubscriptionArn + ')'
```
Lines 116 and 117 are now both indented only two spaces, where they should be indented three. Interesting that our eslint config did not catch this.

gmetzker replied:
done

jthomerson commented on src/index.js:
```diff
@@ -54,67 +56,80 @@ module.exports = Class.extend({
+
+      return this.provider.request('Lambda', 'getFunction', params, this._opts.stage, this._opts.region)
          .then(function(fn) {
+
```
erroneous whitespace that needs to be removed

gmetzker replied:
done

jthomerson commented on src/index.js:
```diff
@@ -125,7 +140,10 @@ module.exports = Class.extend({
             // NOTE: does not support NextToken and paginating through subscriptions at this point
-            return Q.ninvoke(sns, 'listSubscriptionsByTopic', { TopicArn: topicArn });
+            return self.provider.request('SNS', 'listSubscriptionsByTopic',
+              { TopicArn: topicArn }, self._opts.stage, self._opts.region
```
Indentation is incorrect here. Please just follow the same pattern you did in other places where the params are a separate variable so that this all fits on one line. That makes it more consistent with the rest of the code.

gmetzker replied:
done

jthomerson commented on src/index.js:
```diff
@@ -125,7 +140,10 @@ module.exports = Class.extend({
+            return self.provider.request('SNS', 'listSubscriptionsByTopic',
+              { TopicArn: topicArn }, self._opts.stage, self._opts.region
+            );
+
```
erroneous whitespace that needs to be removed

gmetzker replied:
done

jthomerson commented on src/index.js:
```diff
@@ -136,6 +154,8 @@ module.exports = Class.extend({
             };
          });
+
+
```
erroneous whitespace that needs to be removed

gmetzker replied:
done

jthomerson commented on src/tests/index.test.js:
```diff
@@ -1,16 +1,476 @@
 describe('serverless-plugin-external-sns-events', function() {
 
+   function createMockServerless(requestFunc) {
+
```
erroneous whitespace that needs to be removed

gmetzker replied:
done

jthomerson commented on src/tests/index.test.js:
```diff
@@ -1,16 +1,476 @@
+
+      serverless = {
+         getProvider: function(providerName) {
+
```
erroneous whitespace that needs to be removed

gmetzker replied:
done

jthomerson commented on src/tests/index.test.js:
```diff
@@ -1,16 +1,476 @@
+         } }
+         // cli: { log: function(val) {
+         //    process.stdout.write(val + '\n');
+         // } }
```
commented-out code should not be committed

gmetzker replied:
done

jthomerson commented on src/tests/index.test.js:
```diff
@@ -1,16 +1,476 @@
+         },
+         cli: { log: function() {
+            return;
+         } }
```
please add trailing comma (explanation above in other comment)

gmetzker replied:
done

jthomerson commented on src/tests/index.test.js:
```diff
@@ -1,16 +1,476 @@
+      };
+
+      return serverless;
+
```
erroneous whitespace that needs to be removed

gmetzker replied:
done

jthomerson commented on src/tests/index.test.js:
```diff
@@ -1,16 +1,476 @@
+   }
+
+   function createMockRequest(requestStub) {
+
```
erroneous whitespace that needs to be removed

gmetzker replied:
done

jthomerson commented on src/tests/index.test.js:
```diff
@@ -1,16 +1,476 @@
+   function createMockRequest(requestStub) {
+
+      return function() {
+
```
erroneous whitespace that needs to be removed

gmetzker replied:
done

jthomerson commented on src/tests/index.test.js:
```diff
@@ -1,16 +1,476 @@
+
+         var reqArgs = Array.prototype.slice.call(arguments);
+
+
```
erroneous whitespace that needs to be removed (remove one line, leave the other)

gmetzker replied:
done

jthomerson commented on src/tests/index.test.js:
```diff
@@ -1,16 +1,476 @@
+            }
+
+            reject(new Error('Call to request() with unexpected arguments:  ' +
+            JSON.stringify(reqArgs)));
```
Indentation doesn't match standards. I'm again confused why eslint didn't catch this. Please reformat.

gmetzker replied:
Yeah this one was tripping up your eslint config.  I tried indenting several different ways but it complains either way.  I'll just put it on one line.

jthomerson commented on src/tests/index.test.js:
```diff
@@ -1,16 +1,476 @@
+      it('can compile lambda permission with correct FunctionName and SourceArn', function() {
+
+         // ARRANGE:
+
```
I'm going to stop commenting on all the extra blank lines. Please just remove blank lines where they don't match the pattern you observe in the rest of the code. In general, code should read like paragraphs - a thought could be grouped into one paragraph, and different thoughts should be separated with a blank line between them.

gmetzker replied:
done

jthomerson commented on src/tests/index.test.js:
```diff
@@ -1,16 +1,476 @@
-      // TODO: write tests
+      it('can compile lambda permission with correct FunctionName and SourceArn', function() {
+
+         // ARRANGE:
```
The "ARRANGE", "ACT", "ASSERT" comments don't really add much. The purpose of a comment should be to explain what's happening, but this needs no explanation - at least not a simple one that's merely defining the three stages inherent to every unit test.

gmetzker replied:
I'll go ahead and remove these.  Sorry this is just a style I always follow.  

I've always found it clearly lays out the different sections of a test.  I do find it's helpful visual book mark especially to find what is under test / action.

jthomerson commented on src/tests/index.test.js:
```diff
@@ -1,16 +1,476 @@
+         mockServerless = createMockServerless(createMockRequest(sinon.stub()));
+         spyRequestFunc = sinon.spy(mockServerless.getProvider('aws'), 'request');
+
+         plugin = new Plugin(mockServerless, { });
```
we initialize empty objects with no space in them, e.g. `{}`

gmetzker replied:
done

jthomerson commented on src/tests/index.test.js:
```diff
@@ -1,16 +1,476 @@
+      it('can return SNS Subscription info when subscription exists', function() {
+
+         // ARRANGE:
+         var mockServerless, requestMethod, actual, requestStub, plugin, lambdaArn, topicArn,
```
uninitialized variables go last in our var declarations

gmetzker replied:
done

gmetzker replied:
done

jthomerson commented on src/tests/index.test.js:
```diff
@@ -1,16 +1,476 @@
+      it('can return undefined Subscription info when subscription does NOT exist', function() {
+
+         // ARRANGE:
+         var mockServerless, requestMethod, actual, requestStub, plugin, lambdaArn, topicArn,
```
same here - I won't comment on it each time it appears - please fix them all

gmetzker replied:
done

jthomerson reviewed (changes requested):
This is some really great work! I apologize greatly for the delay in reviewing this. I was away for five weeks and you submitted this on my last week away. I've been buried since then, but now am getting back caught up on things. I left a bunch of comments that are mostly about cosmetic coding standards issues. The one more major issue is the addition of the Bluebird dependency. Please replace that with `q` and update the tests. I love the fact that you were able to remove the clunky dependency on the AWS SDK, and with it the runtime dependency on a promise library (needing `q` now only for the tests). And we *absolutely love* all the work that went into increasing test coverage.

One final thing is to try to clean up the commits by rebasing and squashing most of them. We always try to keep a clean commit history so that later `git blame` is more useful. So, we don't tons of individual commits for what is one unit of work, and we don't do "fixed code review things" commits. So, after you make your changes could you rebase this down into one or two or three commits. Each commit should be about a specific thing. For example, in this PR it seems like you could have:

1. added test coverage
2. changed plugin to use built-in AWS provider
3. fixed lambda event permissions so that the subscription will appear in the AWS console

That makes a much more useful git history.

Thanks so much! Let me know if you have any questions and I'll respond promptly.

gmetzker commented:
I'm happy to address these but it's difficult to anticipate style issues that aren't part of the lint config.

Any ETA when your lint config will be more lined up with your review requests?

Gary

gmetzker commented:
I believe I have addressed all the review items, and the commits are squashed.

If you have additional whitespace requests I would ask if you could publish and updated version of your lint configuration that aligns more closely with your expectations.  

The `package.json` is currently using `"eslint-config-silvermine": "1.0.0"`

Thanks.

jthomerson commented:
@gmetzker Thanks so much! Yesterday I created #10 to look into why eslint wasn't catching those indentation and other styling issues, as it's our goal that it will catch 95% of all styling issues. Some things, like the additional whitespace lines, are not things that are easily enforced by a set of automated rules because they're more subjective. We hope that developers will style their code consistently with the surrounding code.

Thanks again so much for all the awesome work on this. I'll start right now on the re-review.

gmetzker commented:
Sorry, I had to push to my branch again because I found commit history had an incorrect email.
Let me know if there is anything else I can address.  

BTW, Here are a few eslint rules that might align with your style. I tend to use `airbnb-base` which is quite popular and has a lot of these built in.

- `comma-dangle`:  http://eslint.org/docs/rules/comma-dangle
- `padded-blocks`:  http://eslint.org/docs/rules/padded-blocks

jthomerson commented:
Great work @gmetzker! I made some minor styling adjustments and merged it via #11. Release 1.1.0 is cut and published [to npm](https://www.npmjs.com/package/serverless-plugin-external-sns-events). Give it a whirl and let us know if there are any other issues.

Thanks for the great work on this!

gmetzker commented:
Seems like it's working good.  Thanks for the pull!

Govinda-Fichtner commented:
@jthomerson @gmetzker It is really great that this is finally merged. As long as standard serverless is not supporting external SNS topics out of the box, this is a really important feature for me! Great work, guys!
